consensus/src/consensus/mod.rs

```diff
@@ -767,11 +770,82 @@ impl ConsensusApi for Consensus {
+        &self,
+        accepting_block: Hash,
+        tx_ids: Option<Vec<TransactionId>>,
+        tx_type: TransactionType,
+    ) -> ConsensusResult<TransactionQueryResult> {
+        // We need consistency between the acceptance store and the block transaction store,
         let _guard = self.pruning_lock.blocking_read();
-        self.virtual_processor.get_populated_transaction(txid, accepting_block_daa_score, self.get_retention_period_root())
+
+        match tx_type {
+            TransactionType::Transaction => {
+                if let Some(tx_ids) = tx_ids {
+                    let tx_id_set = HashSet::<TransactionId>::from_iter(tx_ids);
+
+                    Ok(TransactionQueryResult::Transaction(Arc::new(
+                        self.acceptance_data_store
+                            .get(accepting_block)
+                            .map_err(|_| ConsensusError::MissingData(accepting_block))?
+                            .unwrap_or_clone()
+                            .into_iter()
+                            .flat_map(|mbad| {
+                                self.get_block_transactions(
+                                    mbad.block_hash,
+                                    Some(
+                                        mbad.accepted_transactions
+                                            .into_iter()
+                                            .filter_map(|atx| {
+                                                if tx_id_set.contains(&atx.transaction_id) {
+                                                    Some(atx.index_within_block)
+                                                } else {
+                                                    None
+                                                }
+                                            })
+                                            .collect(),
+                                    ),
+                                )
+                            })
+                            .flatten()
+                            .collect::<Vec<_>>(),
+                    )))
+                } else {
+                    Ok(TransactionQueryResult::Transaction(Arc::new(
+                        self.acceptance_data_store
+                            .get(accepting_block)
+                            .map_err(|_| ConsensusError::MissingData(accepting_block))?
+                            .unwrap_or_clone()
+                            .into_iter()
+                            .flat_map(|mbad| {
+                                self.get_block_transactions(
+                                    mbad.block_hash,
+                                    Some(mbad.accepted_transactions.iter().map(|atx| atx.index_within_block).collect()),
+                                )
+                            })
+                            .flatten()
+                            .collect::<Vec<_>>(),
+                    )))
+                }
+            }
```

Comment:
factorable, but has it been made for readability reasons?

if so, should it deserve two functions instead?